File: infrastructure/services/iam-service.ts
import * as AWS from 'aws-sdk';

export interface IAMServiceConfig {
  projectName: string;
  region?: string;
}

export class IAMService {
  private iam: AWS.IAM;
  private sts: AWS.STS;
  private projectName: string;
  private region: string;
  
  public executionRoleArn: string;
  public taskRoleArn: string;
  public codeBuildRoleArn: string;

  constructor(private config: IAMServiceConfig) {
    this.projectName = config.projectName;
    this.region = config.region || 'us-east-1';
    
    AWS.config.update({ region: this.region });
    this.iam = new AWS.IAM();
    this.sts = new AWS.STS();
  }

  async initialize(): Promise<void> {
    // Create IAM roles
    this.taskRoleArn = await this.createOrUpdateTaskRole();
    this.executionRoleArn = await this.createOrUpdateExecutionRole();
    this.codeBuildRoleArn = await this.createOrUpdateCodeBuildRole();
  }

  private async createOrUpdateTaskRole(): Promise<string> {
    const roleName = `${this.projectName}-task-role`;
    
    try {
      // Check if role already exists
      const role = await this.iam.getRole({ RoleName: roleName }).promise();
      console.log(`Found existing task role: ${roleName}`);
      
      // Update the inline policy for ECS exec access
      await this.attachECSExecPolicy(roleName);
      
      return role.Role.Arn!;
    } catch (error: any) {
      if (error.code === 'NoSuchEntity') {
        console.log(`Creating new task role: ${roleName}`);
        
        // Create the role
        const roleResult = await this.iam.createRole({
          RoleName: roleName,
          AssumeRolePolicyDocument: JSON.stringify({
            Version: '2012-10-17',
            Statement: [{
              Action: 'sts:AssumeRole',
              Effect: 'Allow',
              Principal: { Service: 'ecs-tasks.amazonaws.com' }
            }]
          }),
          Tags: [{ Key: 'Name', Value: roleName }]
        }).promise();

        // Attach ECS exec policy
        await this.attachECSExecPolicy(roleName);

        console.log(`Created task role: ${roleResult.Role.Arn}`);
        return roleResult.Role.Arn!;
      } else {
        throw error;
      }
    }
  }

  private async attachECSExecPolicy(roleName: string): Promise<void> {
    const policyName = `${this.projectName}-ecs-exec-policy`;
    
    const ecsExecPolicy = {
      Version: '2012-10-17',
      Statement: [
        {
          Effect: 'Allow',
          Action: [
            'ssmmessages:CreateControlChannel',
            'ssmmessages:CreateDataChannel',
            'ssmmessages:OpenControlChannel',
            'ssmmessages:OpenDataChannel'
          ],
          Resource: '*'
        }
      ]
    };

    try {
      // Try to update existing policy
      await this.iam.putRolePolicy({
        RoleName: roleName,
        PolicyName: policyName,
        PolicyDocument: JSON.stringify(ecsExecPolicy)
      }).promise();
      
      console.log(`Updated ECS exec policy for role: ${roleName}`);
    } catch (error) {
      console.error(`Failed to attach ECS exec policy: ${error}`);
      throw error;
    }
  }

  private async createOrUpdateExecutionRole(): Promise<string> {
    const roleName = `${this.projectName}-execution-role`;
    const accountId = await this.getAccountId();
    
    try {
      // Check if role already exists
      const role = await this.iam.getRole({ RoleName: roleName }).promise();
      console.log(`Found existing execution role: ${roleName}`);
      
      // Update the inline policy for Secrets Manager access
      await this.attachSecretsManagerPolicy(roleName, accountId);
      
      return role.Role.Arn!;
    } catch (error: any) {
      if (error.code === 'NoSuchEntity') {
        console.log(`Creating new execution role: ${roleName}`);
        
        // Create the role
        const roleResult = await this.iam.createRole({
          RoleName: roleName,
          AssumeRolePolicyDocument: JSON.stringify({
            Version: '2012-10-17',
            Statement: [{
              Action: 'sts:AssumeRole',
              Effect: 'Allow',
              Principal: { Service: 'ecs-tasks.amazonaws.com' }
            }]
          }),
          Tags: [{ Key: 'Name', Value: roleName }]
        }).promise();

        // Attach the basic ECS task execution policy
        await this.iam.attachRolePolicy({
          RoleName: roleName,
          PolicyArn: 'arn:aws:iam::aws:policy/service-role/AmazonECSTaskExecutionRolePolicy'
        }).promise();

        // Attach Secrets Manager policy
        await this.attachSecretsManagerPolicy(roleName, accountId);

        console.log(`Created execution role: ${roleResult.Role.Arn}`);
        return roleResult.Role.Arn!;
      } else {
        throw error;
      }
    }
  }

  private async attachSecretsManagerPolicy(roleName: string, accountId: string): Promise<void> {
    const policyName = `${this.projectName}-secrets-policy`;
    
    const secretsPolicy = {
      Version: '2012-10-17',
      Statement: [
        {
          Effect: 'Allow',
          Action: [
            'secretsmanager:GetSecretValue'
          ],
          Resource: [
            `arn:aws:secretsmanager:${this.region}:${accountId}:secret:${this.projectName}/*`
          ]
        }
      ]
    };

    try {
      // Try to update existing policy
      await this.iam.putRolePolicy({
        RoleName: roleName,
        PolicyName: policyName,
        PolicyDocument: JSON.stringify(secretsPolicy)
      }).promise();
      
      console.log(`Updated Secrets Manager policy for role: ${roleName}`);
    } catch (error) {
      console.error(`Failed to attach Secrets Manager policy: ${error}`);
      throw error;
    }
  }

  private async createOrUpdateCodeBuildRole(): Promise<string> {
    const roleName = `${this.projectName}-codebuild-role`;
    
    try {
      // Check if role already exists
      const role = await this.iam.getRole({ RoleName: roleName }).promise();
      console.log(`Found existing CodeBuild role: ${roleName}`);
      return role.Role.Arn!;
    } catch (error: any) {
      if (error.code === 'NoSuchEntity') {
        console.log(`Creating new CodeBuild role: ${roleName}`);
        
        const assumeRolePolicy = {
          Version: '2012-10-17',
          Statement: [
            {
              Action: 'sts:AssumeRole',
              Effect: 'Allow',
              Principal: {
                Service: 'codebuild.amazonaws.com',
              },
            },
          ],
        };

        const inlinePolicy = {
          Version: '2012-10-17',
          Statement: [
            {
              Effect: 'Allow',
              Action: [
                'logs:CreateLogGroup',
                'logs:CreateLogStream',
                'logs:PutLogEvents',
              ],
              Resource: 'arn:aws:logs:*:*:*',
            },
            {
              Effect: 'Allow',
              Action: [
                'ecr:BatchCheckLayerAvailability',
                'ecr:GetDownloadUrlForLayer',
                'ecr:BatchGetImage',
                'ecr:GetAuthorizationToken',
              ],
              Resource: '*',
            },
            {
              Effect: 'Allow',
              Action: [
                'ecs:UpdateService',
                'ecs:DescribeServices',
                'ecs:DescribeTaskDefinition',
                'ecs:RegisterTaskDefinition',
              ],
              Resource: '*',
            },
            {
              Effect: 'Allow',
              Action: ['iam:PassRole'],
              Resource: '*',
            },
          ],
        };

        const role = await this.iam.createRole({
          RoleName: roleName,
          AssumeRolePolicyDocument: JSON.stringify(assumeRolePolicy),
          InlinePolicies: [
            {
              PolicyName: `${this.projectName}-codebuild-policy`,
              PolicyDocument: JSON.stringify(inlinePolicy),
            },
          ],
          Tags: [{ Key: 'Name', Value: roleName }]
        }).promise();

        // Attach managed policy
        await this.iam.attachRolePolicy({
          RoleName: roleName,
          PolicyArn: 'arn:aws:iam::aws:policy/AWSCodeBuildDeveloperAccess'
        }).promise();

        console.log(`Created CodeBuild role: ${role.Role.Arn}`);
        return role.Role.Arn!;
      } else {
        throw error;
      }
    }
  }

  private async getAccountId(): Promise<string> {
    const identity = await this.sts.getCallerIdentity().promise();
    return identity.Account!;
  }
}

File: backend/infrastructure/services/iam_service.py
import json
import logging
from typing import Tuple
from utils.aws_client import create_client

logger = logging.getLogger(__name__)


class IAMService:
    def __init__(
        self, project_name: str, region: str = "us-east-1", aws_credentials=None
    ):
        self.project_name = project_name
        self.region = region
        self.aws_credentials = aws_credentials

        # Initialize AWS clients with custom credentials if provided
        self.iam = create_client("iam", region, aws_credentials)
        self.sts = create_client("sts", region, aws_credentials)

        self.execution_role_arn: str = ""
        self.task_role_arn: str = ""
        self.codebuild_role_arn: str = ""

    async def initialize(self):
        """Initialize IAM roles"""
        self.execution_role_arn = await self._create_execution_role()
        self.task_role_arn = await self._create_task_role()
        self.codebuild_role_arn = await self._create_codebuild_role()

    async def _get_account_id(self) -> str:
        """Get AWS account ID"""
        response = self.sts.get_caller_identity()
        return response["Account"]

    async def _create_execution_role(self) -> str:
        """Create ECS task execution role"""
        role_name = f"{self.project_name}-ecs-execution-role"

        try:
            # Check if role exists
            response = self.iam.get_role(RoleName=role_name)
            logger.info(f"Found existing execution role: {response['Role']['Arn']}")
            return response["Role"]["Arn"]
        except self.iam.exceptions.NoSuchEntityException:
            pass

        # Create new role
        trust_policy = {
            "Version": "2012-10-17",
            "Statement": [
                {
                    "Effect": "Allow",
                    "Principal": {"Service": "ecs-tasks.amazonaws.com"},
                    "Action": "sts:AssumeRole",
                }
            ],
        }

        response = self.iam.create_role(
            RoleName=role_name,
            AssumeRolePolicyDocument=json.dumps(trust_policy),
            Tags=[{"Key": "Name", "Value": role_name}],
        )

        role_arn = response["Role"]["Arn"]

        # Attach policies
        policies = [
            "arn:aws:iam::aws:policy/service-role/AmazonECSTaskExecutionRolePolicy"
        ]

        for policy_arn in policies:
            self.iam.attach_role_policy(RoleName=role_name, PolicyArn=policy_arn)

        # Create and attach inline policy for Secrets Manager
        secrets_policy = {
            "Version": "2012-10-17",
            "Statement": [
                {
                    "Effect": "Allow",
                    "Action": [
                        "secretsmanager:GetSecretValue",
                        "secretsmanager:DescribeSecret",
                    ],
                    "Resource": f"arn:aws:secretsmanager:{self.region}:*:secret:autoform/*",
                }
            ],
        }

        self.iam.put_role_policy(
            RoleName=role_name,
            PolicyName=f"{self.project_name}-secrets-access",
            PolicyDocument=json.dumps(secrets_policy),
        )

        logger.info(f"Created execution role: {role_arn}")
        return role_arn

    async def _create_task_role(self) -> str:
        """Create ECS task role"""
        role_name = f"{self.project_name}-ecs-task-role"

        try:
            # Check if role exists
            response = self.iam.get_role(RoleName=role_name)
            logger.info(f"Found existing task role: {response['Role']['Arn']}")
            return response["Role"]["Arn"]
        except self.iam.exceptions.NoSuchEntityException:
            pass

        # Create new role
        trust_policy = {
            "Version": "2012-10-17",
            "Statement": [
                {
                    "Effect": "Allow",
                    "Principal": {"Service": "ecs-tasks.amazonaws.com"},
                    "Action": "sts:AssumeRole",
                }
            ],
        }

        response = self.iam.create_role(
            RoleName=role_name,
            AssumeRolePolicyDocument=json.dumps(trust_policy),
            Tags=[{"Key": "Name", "Value": role_name}],
        )

        role_arn = response["Role"]["Arn"]

        # Add policies for ECS Exec
        exec_policy = {
            "Version": "2012-10-17",
            "Statement": [
                {
                    "Effect": "Allow",
                    "Action": [
                        "ssmmessages:CreateControlChannel",
                        "ssmmessages:CreateDataChannel",
                        "ssmmessages:OpenControlChannel",
                        "ssmmessages:OpenDataChannel",
                    ],
                    "Resource": "*",
                }
            ],
        }

        self.iam.put_role_policy(
            RoleName=role_name,
            PolicyName=f"{self.project_name}-ecs-exec",
            PolicyDocument=json.dumps(exec_policy),
        )

        logger.info(f"Created task role: {role_arn}")
        return role_arn

    async def _create_codebuild_role(self) -> str:
        """Create CodeBuild service role"""
        role_name = f"{self.project_name}-codebuild-role"
        account_id = await self._get_account_id()

        try:
            # Check if role exists
            response = self.iam.get_role(RoleName=role_name)
            logger.info(f"Found existing CodeBuild role: {response['Role']['Arn']}")
            return response["Role"]["Arn"]
        except self.iam.exceptions.NoSuchEntityException:
            pass

        # Create new role
        trust_policy = {
            "Version": "2012-10-17",
            "Statement": [
                {
                    "Effect": "Allow",
                    "Principal": {"Service": "codebuild.amazonaws.com"},
                    "Action": "sts:AssumeRole",
                }
            ],
        }

        response = self.iam.create_role(
            RoleName=role_name,
            AssumeRolePolicyDocument=json.dumps(trust_policy),
            Tags=[{"Key": "Name", "Value": role_name}],
        )

        role_arn = response["Role"]["Arn"]

        # Create and attach inline policy
        codebuild_policy = {
            "Version": "2012-10-17",
            "Statement": [
                {
                    "Effect": "Allow",
                    "Action": [
                        "logs:CreateLogGroup",
                        "logs:CreateLogStream",
                        "logs:PutLogEvents",
                    ],
                    "Resource": f"arn:aws:logs:{self.region}:{account_id}:log-group:/aws/codebuild/*",
                },
                {
                    "Effect": "Allow",
                    "Action": ["s3:GetObject", "s3:PutObject"],
                    "Resource": [
                        f"arn:aws:s3:::{self.project_name.lower()}-builds-{self.region}/*",
                        f"arn:aws:s3:::codebuild-{self.region}-{account_id}-*/*",
                        # Add wildcard pattern to handle various project name formats
                        f"arn:aws:s3:::*-builds-{self.region}/*"
                    ],
                },
                {
                    "Effect": "Allow",
                    "Action": ["s3:ListBucket"],
                    "Resource": [
                        f"arn:aws:s3:::{self.project_name.lower()}-builds-{self.region}",
                        f"arn:aws:s3:::codebuild-{self.region}-{account_id}-*",
                        # Add wildcard pattern to handle various project name formats
                        f"arn:aws:s3:::*-builds-{self.region}"
                    ],
                },
                {
                    "Effect": "Allow",
                    "Action": [
                        "ecr:GetAuthorizationToken",
                        "ecr:BatchCheckLayerAvailability",
                        "ecr:GetDownloadUrlForLayer",
                        "ecr:BatchGetImage",
                        "ecr:PutImage",
                        "ecr:InitiateLayerUpload",
                        "ecr:UploadLayerPart",
                        "ecr:CompleteLayerUpload",
                    ],
                    "Resource": "*",
                },
            ],
        }

        self.iam.put_role_policy(
            RoleName=role_name,
            PolicyName=f"{self.project_name}-codebuild-policy",
            PolicyDocument=json.dumps(codebuild_policy),
        )

        logger.info(f"Created CodeBuild role: {role_arn}")
        return role_arn
